feat: added toc dropup like in Kiwix-PWA

heropj opened the pull request:
fix #1212 

- Implemented a fully functional TOC drop-up, inspired by Kiwix-PWA and PR #1241


- When user clicks on drop-up button, TOC List is set up and displayed as drop-up menu.
- It is closed when drop-up button loses focus or when a section is clicked.
- Clicking a section scrolls the view to that section.

![image](https://github.com/user-attachments/assets/75dec5ce-3d94-4f90-86df-7c8c5c1a5a48)

Jaifroid commented:
@heropj Many thanks for this PR. I'm just running tests, and then will do some manual testing.

Jaifroid commented:
Thanks for this PR. I've done some manual testing, and I noticed these issues:

* It works in Wikipedia, but the ToC closes as soon as the user jumps to a heading. It should stay open until the user clicks out of the ToC, in case the user wants to jump around i the article. Alternatively, you should make sure that the header and footer remain visible after the jump, so that users can easily re-open the ToC. Otherwise it's difficult, ergonomically, to re-open it after a jump.
* I tried it with a Gutenberg ZIM, and although the headings showed, I got a console error when clicking on one of the headings. After that error, the ToC wouldn't close any more. See screenshot.

![image](https://github.com/user-attachments/assets/0fcc4ca0-09a5-4343-b161-a8889d0928a4)

Jaifroid reviewed (changes requested):
Please see previous comment for some issue you need to address. Please test the UI fully with different ZIM types and focus on useability. Thanks!

heropj commented:
hello @Jaifroid, Thank-you for reviewing and highlighting these issues.

fix:

- For the first ux related issue, The TOC will now close only when user clicks outside of it, For this I have added 'click' event listener on the iframe, which is applicable as the iframe content is same-origin with respect to the parent page. I found that scrolling down and then up, to make TOC and Header visible after a jump, inconsistent many times. Hence, I think keeping the TOC on-screen until the user clicks out of it is a better choice. Also, the background changes for a short amount of time to highlight the section the user has scrolled to.
- The second issue occurred because the Gutenberg ZIM didn't have an ```id``` set to its headings, it worked for Wikipedia because all heading tags in Wikipedia had a unique id. To tackle this, I have modified the ```getHeadingObjects``` function. It now generates a unique id, if it doesn't already exist, and assigns it to the fetched headings.

Jaifroid commented:
Thank you, will test!

Jaifroid reviewed (changes requested):
Thank you very much, functionality is now as expected!
As app.js is now very large, it would be better to move as much of the new code in app.js as possible to uiUtil.js. I think pretty much all of it could move and just be called as necessary from app.js. You should have a `setUpToC` function in uiUtil.js that would simply be called from app.js where your current comment is for "Setup table of contents...".

Instead of setting the `iframe` variable in uiUtil.js, try using the existing `articleContainer` variable in uiUtil.js. Any other variables you need can be passed to the more specific functions.

Jaifroid commented:
@heropj Apologies for the long delay in reviewing this -- I had some very intensive tasks to complete at work and lost sight of this. The PR is very good, and nearly there!

Jaifroid commented:
NB Please update this branch from main, and also be sure to run `npm install` again in the root of this repo in your fork, because some of the development tools, including Vite.js, have been updated with security fixes.

heropj commented:
hello @Jaifroid, Pls check, I have moved the code to ```uiUtil.js```

Jaifroid commented:
@heropj Did you manually test your code before requesting review? As you can see, tests have immediately failed, and although I didn't manually test, I'm guessing the app doesn't actually run with the changed code. Please debug and fix!

![image](https://github.com/user-attachments/assets/7aeb627e-1e90-4958-b18d-dcbd7d74a1fb)

Jaifroid commented:
As well as checking out hat issue, please update this branch because some fixes to tests have been made.

heropj commented:
hello @Jaifroid , could you please check why the tests are failing, the code is working fine on my system.., idk why it is failing here.
I reviewed the code and fixed the previous issue (was an easy fix, just have to apply documentloaded event-listener before accessing 'dropup' element.)

This time the issue is related to something else, pls check, I'll try to fix it.

Jaifroid commented:
Re-running tests...

Jaifroid commented:
OK, tests have passed! I'll review the code...

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -172,6 +172,157 @@ function slideAway (e) {
+        if (/^h1$/i.test(heading.tagName)) {
+            dropupHtml += '<li style="font-size:' + 18 + 'px;"><a style="color: black;" href="#" data-heading-id="' + heading.id + '">' + heading.textContent + '</a></li>';
+        } else if (/^h2$/i.test(heading.tagName)) {
+            dropupHtml += '<li style="margin-top:6px;margin-left:6px;font-size:' + 16 + 'px;"><a style="color: black;" href="#" data-heading-id="' + heading.id + '">' + heading.textContent + '</a></li>';
```
It would be much better to define a class for each of these, like `class="toc-item-h1"`, `class="toc-item-h2"`, etc., and put the style rules in `app.css`.

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -172,6 +172,157 @@ function slideAway (e) {
+                sectionEle.style.backgroundColor = '#bdd1e5';
+                setTimeout(function () {
+                    sectionEle.style.backgroundColor = '';
+                }, 400);
```
This timeout is so short, I didn't even realize you had added that feature! I'd make this more like 2000 (2 seconds)!

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -172,6 +172,157 @@ function slideAway (e) {
+        });
+}
+
+document.addEventListener('DOMContentLoaded', function () {
```
Please add a comment above to say what this code block does (I know what it does, but it's a good idea to comment explicitly and label it as part of the ToC setup code!

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -172,6 +172,157 @@ function slideAway (e) {
+    const ToCList = document.getElementById('ToCList');
+
+    // close TOC when 'click' is not on dropup or TOCList
+    articleContainer.addEventListener('load', function () {
```
This code risks setting up multiple event listeners on the iframe. Each time a new document is loaded into the iframe. Because the iframe is static, and not destroyed and reloaded each time... So I think it would be best explicitly to remove the event listener before you re-add it. You do this by making the function into a variable, like `const onContainerLoad = function () {....exisitng code inside function...};`, and place this above the remove and add event listeners (just above this spot is fine). Then do `aricleContainer.removeEventListener('load', onContainerLoad);` followed by `articleContainer.addEventListener('load', onContainerLoad);`

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -172,6 +172,157 @@ function slideAway (e) {
+// Inject table of contents list into dropup element and scroll selection into view
+function setUpTOC () {
+    // var iframe = document.getElementById('articleContent');
+    var innerDoc = articleContainer.contentDocument;
```
Also, I notice that below you use the fallback `articleContainer.contentDocument || articleContainer.contentWindow.document`. You should use the same here (but you'll need to put it in parentheses).

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -172,6 +172,157 @@ function slideAway (e) {
+            closeTOC(); // Close TOC when clicking inside the iframe
+        });
+    });
+    document.addEventListener('click', function (event) {
+        if (!dropup.contains(event.target) && !ToCList.contains(event.target)) {
+            closeTOC();
+        }
+    });
+    dropup.addEventListener('click', function () {
+        const isVisible = getComputedStyle(ToCList).display !== 'none';
+        if (isVisible) {
+            ToCList.style.display = 'none';
+        } else {
+            setUpTOC();
+            ToCList.style.display = 'flex';
+            ToCList.style.flexDirection = 'column';
+        }
+    });
```
Why do you add the same event listener twice to the dropup? Consolidate into single function. Also, the dropup is a static element, so you'll be adding multiple event listeners each time a new article is loaded. So, you'll need to do the same `removeEventListener` as above first.

heropj replied:
hello @Jaifroid , This was done to handle click event on dropup separately than on other parts of the document.
The first one is for the whole document object, to handle clicks outside of the TOCList (closes the toc).
The second one is added on the dropup button, which handles toggling of TOCList, not only closing.
I tried to use the 'click' event of document for the dropup, but it was causing some issues..hence decided to go with seperate event listener for dropup button.

Jaifroid reviewed (changes requested):
Thanks, it's looking good and works well in manual tests -- just a few code changes. Additionally, you're mixing up the `var` style and the `const` / `let` style. I suggest you modernize and use the latter throughout your code for this PR, as it's good practice.

Jaifroid commented:
Also, please update from main, and when you pull the changes to your local repo, be sure to run `npm install` again (because quite a lot of packages have changed).

heropj commented:
hello @Jaifroid, thank you for the review. I was very busy with a family function last week, I'll make the changes by tomorrow. 🚀

Jaifroid reviewed (changes requested):
Nearly there! Just some very minor things and ensuring compatibility with old platforms.

Jaifroid commented:
Also please update the branch.

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -171,6 +171,174 @@ function slideAway (e) {
     }
 }
 
+/*
+  * Returns a list of headings from an article
+    * @param {String} the page for which table of cotents needs to be listed
+    * @returns {List} a list of all headings as objects
+*/
```
```suggestion
/**
 * Extracts a list of headings from an article and provides methods to interact with them.
 * @class
 * @param {Document} articleDoc The document object of the article from which headings are to be extracted.
 */
```
Improperly formatted JSDoc.

Jaifroid commented on www/js/lib/uiUtil.js:
```diff
@@ -171,6 +171,174 @@ function slideAway (e) {
+    };
+}
+
+/*
+    * Setup table of contents and display the list when the dropup button is clicked
+    * @param {List} a list of all headings as objects
+    * @returns {void}
+*/
+// Inject table of contents list into dropup element and scroll selection into view
```
Ditto. Please correct this header. You list a parameter which isn't in the function declaration, for example. Incorporate the comment "// Inject table of contents..." into the JSDoc declaration, rather than having it afterwards. Format correctly!

Jaifroid commented:
@heropj Are you interested in finishing this PR? It was pretty much ready, minus a few small cleanup issues I identified above.

Jaifroid commented:
OK, no problem, good luck! 🍀

Jaifroid commented:
Just running tests.

Jaifroid reviewed (approved):
Thank you for this very useful PR @heropj ! 🙏🙏 Tested and working in IE11 even, so it's good to merge when tests have passed. Let me know if you're ready on your side, and I'll do the squash/merge, so we can get this into the next release of the app.
